streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java

```diff
@@ -108,6 +124,48 @@ private void unlock() {
         version.topologyLock.unlock();
     }
 
+    public Collection<String> sourceTopologies(final String name) {
+        return builders.get(name).sourceTopicCollection();
+    }
+
+    public boolean needsUpdate(final String threadName) {
+        return threadVersions.get(threadName) < topologyVersion();
+    }
+
+    public void registerThread(final String threadName) {
+        threadVersions.put(threadName, 0L);
+    }
+
+    public void unregisterThread(final String threadName) {
+        threadVersions.remove(threadName);
+    }
+
+
+    public boolean reachedLatestVersion(final String threadName) {
+        boolean rebalance = false;
+        try {
+            lock();
+            final Iterator<TopologyVersionWaiters> iterator = version.activeTopologyWaiters.listIterator();
+            TopologyVersionWaiters topologyVersionWaiters;
+            threadVersions.put(threadName, topologyVersion());
+            while (iterator.hasNext()) {
+                topologyVersionWaiters = iterator.next();
+                final long verison = topologyVersionWaiters.topologyVersion;
+                if (verison <= threadVersions.get(threadName)) {
+                    if (threadVersions.values().stream().allMatch(t -> t >= verison)) {
+                        topologyVersionWaiters.future.complete(null);
+                        iterator.remove();
+                        log.info("thread {} is now on on version {}", threadName, topologyVersionWaiters.topologyVersion);
+                        rebalance = true;
```

Comment:
This is a little tricky, but I think it might make sense to do the reverse. It has to do with how the thread handles being assigned tasks that it doesn't recognize because it hasn't yet updated its topology, which is to just make a note of the assigned task name and then stash it away until we see a topology update, during which we then check if the task was part of this new topology and if so we go on to create it. Which is a long way of saying, it's better to rebalance earlier than later, ie to rebalance when the _first_ thread processes a topology update rather than when the _last_ thread has processed the update. Partly so that the first thread can go ahead and start processing tasks from that topology right away, and partly because if we wait until the last thread acks the topology update then we could get in trouble and end up never rebalancing in extreme cases, for example if a thread gets stuck in processing (like if the user has bad custom logic that sends it into an infinite loop, which is a real example from an escalation I once had)

Happy to chat over Zoom if you want to dig into this a bit more, I thought I had left a more specific TODO comment about when we do/don't want to rebalance but I just read it again and it's wildly unhelpful 😅

Reply:
How about we just take the rebalances for now? We can improve this later

Reply:
You mean just always trigger a rebalance? That SGTM